From 6bfc54ee5d503b9a8e62ec05d0397eb781062104 Mon Sep 17 00:00:00 2001
From: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>
Date: Fri, 20 Oct 2017 08:42:59 +0200
Subject: [PATCH] ov5640: Fix exposure and white balance

Wait hard 750 ms, otherwise the automatic exposure on the camera board does not
start.

Change values of the white balance to get a better hue.

Signed-off-by: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>

---
 drivers/media/platform/mxc/capture/ov5640_mipi.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/media/platform/mxc/capture/ov5640_mipi.c b/drivers/media/platform/mxc/capture/ov5640_mipi.c
index e7ef3ae7cd26..f14dd953a004 100644
--- a/drivers/media/platform/mxc/capture/ov5640_mipi.c
+++ b/drivers/media/platform/mxc/capture/ov5640_mipi.c
@@ -745,9 +745,9 @@ static struct reg_value ov5640_init_setting_30fps_VGA[] = {
 	{0x5000, 0xa7, 0, 0}, {0x5001, 0xa3, 0, 0}, {0x5180, 0xff, 0, 0},
 	{0x5181, 0xf2, 0, 0}, {0x5182, 0x00, 0, 0}, {0x5183, 0x14, 0, 0},
 	{0x5184, 0x25, 0, 0}, {0x5185, 0x24, 0, 0}, {0x5186, 0x09, 0, 0},
-	{0x5187, 0x09, 0, 0}, {0x5188, 0x09, 0, 0}, {0x5189, 0x88, 0, 0},
-	{0x518a, 0x54, 0, 0}, {0x518b, 0xee, 0, 0}, {0x518c, 0xb2, 0, 0},
-	{0x518d, 0x50, 0, 0}, {0x518e, 0x34, 0, 0}, {0x518f, 0x6b, 0, 0},
+	{0x5187, 0x09, 0, 0}, {0x5188, 0x09, 0, 0}, {0x5189, 0x75, 0, 0},
+	{0x518a, 0x54, 0, 0}, {0x518b, 0xe0, 0, 0}, {0x518c, 0xb2, 0, 0},
+	{0x518d, 0x42, 0, 0}, {0x518e, 0x3d, 0, 0}, {0x518f, 0x56, 0, 0},
 	{0x5190, 0x46, 0, 0}, {0x5191, 0xf8, 0, 0}, {0x5192, 0x04, 0, 0},
 	{0x5193, 0x70, 0, 0}, {0x5194, 0xf0, 0, 0}, {0x5195, 0xf0, 0, 0},
 	{0x5196, 0x03, 0, 0}, {0x5197, 0x01, 0, 0}, {0x5198, 0x04, 0, 0},
@@ -2781,6 +2781,9 @@ static int ov5640_init_mode(enum ov5640_frame_rate frame_rate,
 		/* dump the first two frames: 1/30*2 */
 		msec_wait4stable = 67;
 	}
+	/* HACK: Wait 750 ms, otherwise the automatic exposure on the camera board
+		 does not start. */
+	msec_wait4stable = 750;
 	msleep(msec_wait4stable);
 
 	if (mipi_csi2_info) {
